# Pipeline for annotating cell types for sc-/sn-RNA-Seq Analysis in 10X Flex Genomics data

## Usage

`run-cell-types-annotation.sh` is designed to be run as if it was called from this module directory even when called from outside of this directory.

Parameters according to the project and analysis strategy will need to be specified in the following scripts:
- `project_parameters.Config.yaml` located at the `root_dir`
- `future_globals_value` is hard-coded in each `run-cell-types-annotation-<method>.R` script for each method. If necessary, user can increase/decrease resources.
- For the `singleR` method, Celldex references are hard-coded in the `run-cell-types-annotation-SingleR.R` script. User should modify the reference to be used according to their experiment.
- `run-cell-types-annotation.sh`: User should comment in/out the script to use based on the selection of the desired methods for cell type annotation.
- `run-merge-cell-types-annotations-all.R`: The user should run this script to merge cell type annotations from SingleR (both broad and fine resolutions). Additionally, if multiple annotation methods are used (current options include SingleR method, a gene marker list, and a reference dataset), this script will merge the annotations accordingly.


We provide a fixed color palette for annotating cell types. Please ensure that the cell type names in your customized cell type marker reference list and/or reference dataset  (if you use any) exactly match those in the `.figures/palettes/cell_types_palette.tsv`. If any cell types are missing, please submit an [issue](https://github.com/stjude-dnb-binfcore/sc-rna-seq-snap/issues) with the list of missing types, and we will add them to the palette. Alternatively, users are welcome to use their own `cell_types_palette.tsv`, as long as the column names (`cell_type_names` and `hex_codes`), file name, and format match those in `./figures/palettes/cell_types_palette.tsv`. We also provide the list of cell types and color palette for the `label.fine` generated by the  `celldex::HumanPrimaryCellAtlasData()`, saved here: `./figures/palettes/HumanPrimaryCellAtlasData_cell_types_with_colors.tsv`.

Moreover, if the list of cell types is too long, the user might encounter the following error: `The viewport may be too small to show this patchwork.`. This happens because the plot’s size exceeds the default rendering area. To resolve this, we recommend increasing the `fig.width` and `fig.height` settings in your plotting function or chunk options. Alternatively, you can temporarily comment out the `ggsave(...)` call to avoid saving a plot that cannot be properly rendered. Adjusting the plot dimensions helps ensure the entire visualization fits within the available viewport. The plots will still appear in the final HTML report, but the PDF report may fail to generate properly. Unfortunately, there’s no current fix for this issue other than skipping the PDF report generation. However, you will still be able to produce all the necessary output files for visualization within the `rshiny-app` module.


### Run module on an interactive session on HPC within the container

To run all of the Rscripts in this module sequentially on an interactive session on HPC, please run the following command from an interactive compute node:

```
bash run-cell-types-annotation.sh
```

### Run module by using lsf on HPC within the container

There is also the option to run a lsf job on the HPC cluster by using the following command on an HPC node:

```
bsub < lsf-script.txt
```


## Folder content
This folder contains a script tasked to identify cell types annotation with SingleR by using a broad and fine resolution labeling across the project. There is also a script to identify and annotate cell types by using a list of gene markers and/or a reference dataset.

## Folder structure 

The structure of this folder is as follows:

```
├── 01-cell-types-annotation-SingleR-broad.Rmd
├── 02-cell-types-annotation-SingleR-fine.Rmd
├── 03-cell-types-annotation-gene-markers.Rmd
├── 04-cell-types-annotation-reference.Rmd
├── lsf_script.txt
├── plots
|   ├── 01_cell_types_annotation_SingleR_broad
|   |   ├── Report_cell_types_annotation_SingleR_broad_<Sys.Date()>.html
|   |   └── Report_cell_types_annotation_SingleR_broad_<Sys.Date()>.pdf
|   ├── 02_cell_types_annotation_SingleR_fine
|   |   ├── Report_cell_types_annotation_SingleR_fine_<Sys.Date()>.html
|   |   └── Report_cell_types_annotation_SingleR_fine_<Sys.Date()>.pdf
|   ├── 03_cell_types_annotation_gene_markers
|   |   ├── Report_cell_types_annotation_gene_markers_<Sys.Date()>.html
|   |   └── Report_cell_types_annotation_gene_markers_<Sys.Date()>.pdf
|   ├── 04_cell_types_annotation_reference
|   |   ├── Report_cell_types_annotation_reference_<Sys.Date()>.html
|   |   └── Report_cell_types_annotation_reference_<Sys.Date()>.pdf
├── README.md
├── results
|   ├── 01_cell_types_annotation_SingleR_broad
|   ├── 02_cell_types_annotation_SingleR_fine
|   ├── 03_cell_types_annotation_gene_markers
|   ├── 04_cell_types_annotation_reference
|   └── cell_types_annotations_all
├── run-cell-types-annotation-gene-markers.R
├── run-cell-types-annotation-reference.R
├── run-cell-types-annotation-SingleR.R
├── run-cell-types-annotation.sh
├── run-merge-cell-types-annotations-all.R
└── util
|   ├── function-calculate-cell-type-signature.R
|___└── function-cell-type-fractions.R
```

